metis-common/src/main/java/eu/europeana/metis/mongo/MongoClientProvider.java

```diff
@@ -78,36 +79,30 @@ public MongoClientProvider(String connectionUri, Function<String, E> exceptionCr
    *
    * @param properties The properties of the Mongo connection. Note that if the passed properties
    * object is changed after calling this method, those changes will not be reflected when calling
-   * @param clientSettingsBuilder The settings to be applied. The default settings will not be used. The
-   * caller can incorporate the default settings by using an client settings builder obtained from {@link
-   * #getDefaultClientSettingsBuilder()}. {@link #createMongoClient()}.
+   * @param clientSettingsBuilder The settings to be applied. The default settings will not be used.
+   * The caller can incorporate the default settings by using an client settings builder obtained
+   * from {@link #getDefaultClientSettingsBuilder()}. {@link #createMongoClient()}.
    */
-  public MongoClientProvider(MongoProperties<E> properties, Builder clientSettingsBuilder) {
+  public MongoClientProvider(MongoProperties<E> properties, Builder clientSettingsBuilder)
+      throws E {
     final ReadPreference readPreference = Optional.ofNullable(properties.getReadPreferenceValue())
         .map(ReadPreferenceValue::getReadPreferenceSupplier).map(Supplier::get)
         .orElse(DEFAULT_READ_PREFERENCE);
     clientSettingsBuilder.readPreference(readPreference);
 
+    final List<ServerAddress> mongoHosts = properties.getMongoHosts();
     final MongoCredential mongoCredential = properties.getMongoCredentials();
     this.authenticationDatabase = Optional.ofNullable(mongoCredential)
         .map(MongoCredential::getSource).orElse(null);
-    clientSettingsBuilder.applyToSslSettings(builder -> builder.enabled(properties.mongoEnableSsl()));
-    if (mongoCredential == null) {
-      this.creator = () -> {
-        final List<ServerAddress> mongoHosts = properties.getMongoHosts();
-        final MongoClientSettings mongoClientSettings = clientSettingsBuilder
-            .applyToClusterSettings(builder -> builder.hosts(mongoHosts)).build();
-        return MongoClients.create(mongoClientSettings);
-      };
-    } else {
-      this.creator = () -> {
-        final List<ServerAddress> mongoHosts = properties.getMongoHosts();
-        final MongoClientSettings mongoClientSettings = clientSettingsBuilder
-            .applyToClusterSettings(builder -> builder.hosts(mongoHosts))
-            .credential(mongoCredential).build();
-        return MongoClients.create(mongoClientSettings);
-      };
+    clientSettingsBuilder
+        .applyToSslSettings(builder -> builder.enabled(properties.mongoEnableSsl()));
+    clientSettingsBuilder.applyToClusterSettings(builder -> builder.hosts(mongoHosts));
+    if (mongoCredential != null) {
+        clientSettingsBuilder.credential(mongoCredential).build();
```

Comment:
I don't think you should call build() here? Just apply the credential? The build is called below.

Reply:
Yes